components/shared/Search.tsx

```diff
@@ -1,33 +1,158 @@
 "use client";
 
+import { Command } from "cmdk";
+import { useState, useEffect } from "react";
+import Link from "next/link";
+
 import searchIcon from "@/public/searchIcon.svg";
 import shortcutIcon from "@/public/shortcutIcon.svg";
 import Image from "next/image";
+import { useSearchParams, useRouter } from "next/navigation";
+
+import { CreateType, Post } from "@prisma/client";
+import { findPosts } from "@/lib/actions/post.actions";
+import urlManager from "@/lib/utils/urlManager";
+import WorkflowIcon from "./icons/WorkflowIcon";
+import ComponentIcon from "./icons/ComponentIcon";
+import KnowledgeIcon from "./icons/KnowledgeIcon";
+
+const Search = () => {
+  const searchParams = useSearchParams();
+  const router = useRouter();
+  const [open, setOpen] = useState(false);
+  const [searchTerm, setSearchTerm] = useState("");
+  const [posts, setPosts] = useState<Post[]>();
+
+  useEffect(() => {
+    const getPosts = async () => {
+      const posts = await findPosts(searchTerm);
+      if (posts) setPosts(posts as Post[]);
+    };
+    getPosts();
+  }, [searchTerm]);
 
-export default function Search() {
-  function handleSearch(term: string) {
-    console.log(term);
-  }
+  useEffect(() => {
+    const timeout = setTimeout(() => {
+      const newParams = urlManager(searchParams.toString(), {
+        page: "1",
+        term: searchTerm,
+      });
+      router.push(`/?${newParams}`);
+    }, 500);
+    return () => clearTimeout(timeout);
+  }, [searchTerm]);
 
+  // Toggle the menu when ⌘K is pressed
+  useEffect(() => {
+    const down = (e: {
+      key: string;
+      metaKey: any;
+      ctrlKey: any;
+      preventDefault: () => void;
+    }) => {
+      if (e.key === "k" && (e.metaKey || e.ctrlKey)) {
+        e.preventDefault();
+        setOpen((open) => !open);
+      }
+    };
+
+    document.addEventListener("keydown", down);
+    return () => document.removeEventListener("keydown", down);
+  }, []);
+
+  const iconMatch = (post: Post) => {
```

Comment:
Also wanted confirmation about whether this is good for icon matching? I do something similar in multiple places it seems, and I wonder if maybe there's a better way to match icons with their text - create a more generic way to do this?

Reply:
I would just use an object. 

But a switch is perfectly fine.

```js
  const iconClasses = {
    className: "text-purple-500",
    size: 18,
  };
  const iconMap = {
    [CreateType.COMPONENT]: <ComponentIcon {...iconClasses} />,
    [CreateType.WORKFLOW]: <WorkflowIcon {...iconClasses} />,
    [CreateType.KNOWLEDGE]: <KnowledgeIcon {...iconClasses} />,
  };
  ```
  
  ```js
    <Command.Item className="hover:bg-black-600 flex cursor-pointer items-center gap-x-2 px-2 py-2 hover:rounded hover:py-2">
                        {iconMap[post.createType]}
                        <span>{post.title}</span>
     </Command.Item>
```

This is easy to expand and doesn't have all the boilerplate. But there's nothing wrong with the switch statement.